examples/avro_generic_producer_example/avro_generic_producer_example.go

```diff
@@ -0,0 +1,96 @@
+// Example function-based Apache Kafka producer
+package main
+
+/**
+ * Copyright 2022 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import (
+	"fmt"
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"github.com/confluentinc/confluent-kafka-go/schemaregistry"
+	"os"
+)
+
+func main() {
+
+	if len(os.Args) != 4 {
+		fmt.Fprintf(os.Stderr, "Usage: %s <broker> <topic>\n",
+			os.Args[0])
+		os.Exit(1)
+	}
+
+	url := os.Args[1]
+	broker := os.Args[2]
+	topic := os.Args[3]
+
+	p, err := kafka.NewProducer(&kafka.ConfigMap{"bootstrap.servers": broker})
+
+	if err != nil {
+		fmt.Printf("Failed to create producer: %s\n", err)
+		os.Exit(1)
+	}
+
+	fmt.Printf("Created Producer %v\n", p)
+
+	ser := schemaregistry.GenericAvroSerializer{}
+	err = ser.Configure(&kafka.ConfigMap{
+		"schema.registry.url":   url,
+		"auto.register.schemas": true,
+	}, false)
+
+	if err != nil {
+		fmt.Printf("Failed to create serializer: %s\n", err)
+		os.Exit(1)
+	}
+
+	fmt.Printf("Created serializer\n")
+
+	// Optional delivery channel, if not specified the Producer object's
+	// .Events channel is used.
+	deliveryChan := make(chan kafka.Event)
+
+	value := MyRecord{
+		ProductName: "Hello!",
+	}
+	payload, err := ser.Serialize(topic, value)
```

Comment:
Oh interesting - you've decoupled the serdes from the produce call. Not how it's done elsewhere, but I feel this approach is ok. The primary downside I can think of is that you don't get the original record in the context of the delivery report (just the serialized payload). The upside is it's much easier to implement.

Reply:
i guess not having generics forces you to do this.

some more disadvantages:
- Explicit serialization/deserialization is more key-strokes for the user.
- When you instantiate the serializer/deserializer the user needs to specify key or value. They then implicitly specify this again when passing the payload into the produce call, that's superfluous, seems not ideal.
- The serialize call is potentially blocking due to the HTTP call, whereas the produce call isn't (or shouldn't be IMO, i'll need to check how that's setup in the golang client case). I think you ideally want the entire produce operation to be async, and to achieve that the user would need to write some additional code. This may not matter in the goroutine paradigm though (as compared to async/await where you typically never want to block a thread).